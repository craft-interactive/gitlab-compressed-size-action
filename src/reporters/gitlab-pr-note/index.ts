import type { Reporter } from "../../types";

export const gitlabPullRequestNote: Reporter = async (results, services) => {
  const { env, gitlab } = services;
  const prefix = `### :package: file-size-change report`;
  const comments = await gitlab.getPullRequestComments({
    mergeRequestIid: env.CI_MERGE_REQUEST_IID,
  });
  const comment = comments.find((comment) => comment.body.startsWith(prefix));
  const details = `${prefix}

| Path | Size | Change  |        |
| ---- | ---- | ------- | ------ |
${results
  .map(({ id, size, change, isBelowThreshold }) => {
    let path = id;
    let emoji = ":white_check_mark:";

    if (id === "summary") {
      path = "**summary**";
      emoji = isBelowThreshold ? ":white_check_mark:" : ":boom:";
    } else {
      const isUnchanged = change.raw === 0;

      if (isBelowThreshold) {
        if (isUnchanged) {
          emoji = ":small_blue_diamond:";
        } else {
          emoji = ":white_check_mark:";
        }
      } else {
        emoji = ":boom:";
      }
    }

    return `| ${path} | ${size.pretty} | ${change.pretty} (${change.percent}%) | ${emoji} |`;
  })
  .join("\n")}
`;

  if (comment) {
    await gitlab.modifyPullRequestComment({
      mergeRequestIid: env.CI_MERGE_REQUEST_IID,
      commentId: comment.id,
      comment: details,
    });
  } else {
    await gitlab.createPullRequestComment({
      mergeRequestIid: env.CI_MERGE_REQUEST_IID,
      comment: details,
    });
  }
};
