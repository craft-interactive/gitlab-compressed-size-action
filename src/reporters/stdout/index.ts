import type { Reporter } from "../../types";

export const stdout: Reporter = async (results) => console.log(results);
