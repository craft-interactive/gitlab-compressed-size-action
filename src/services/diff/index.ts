import { getReadableFileSize } from "../file";

type Diff = ReturnType<typeof createDiff>;
type DiffStatus = "increased" | "decreased" | "unchanged" | "added";

const createDiff = ({
	id,
	sizes,
	threshold,
}: {
	id: string;
	sizes: { current: number; last: number };
	threshold: number | null;
}) => {
	const change = sizes.current - sizes.last;
	const hasChange = change !== 0;
	const hasIncrease = change > 0;
	let isBelowThreshold = true;
	let changeInPercent: string = "0";
	let status: DiffStatus = "unchanged";

	if (sizes.last) {
		if (hasChange) {
			if (hasIncrease) {
				status = "increased";
			} else {
				status = "decreased";
			}
			changeInPercent = Math.abs(
				((sizes.last - sizes.current) / sizes.last) * 100.0
			).toFixed(2);
		}
	} else {
		status = "added";
		changeInPercent = "100";
	}

	if (threshold) {
		isBelowThreshold = sizes.current <= threshold;
	}

	const prefix = hasChange ? (hasIncrease ? "+" : "-") : "";

	return {
		id,
		status,
		change: {
			raw: change,
			pretty: `${prefix}${getReadableFileSize(Math.abs(change))}`,
			percent: parseFloat(changeInPercent),
		},
		size: {
			raw: sizes.current,
			pretty: getReadableFileSize(sizes.current),
		},
		isBelowThreshold,
	};
};

export { Diff, DiffStatus, createDiff };
