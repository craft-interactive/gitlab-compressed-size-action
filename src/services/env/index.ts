import { cleanEnv, num, str } from "envalid";

type Env = {
	CI_API_V4_URL: string;
	CI_PROJECT_ID: number;

	CI_JOB_NAME: string;

	CI_PIPELINE_IID: number;

	CI_MERGE_REQUEST_IID: number;
};

let instance: Env;

const getEnv = () => {
	if (!instance) {
		instance = cleanEnv(process.env, {
			CI_API_V4_URL: str({}),
			CI_PROJECT_ID: num(),

			CI_JOB_NAME: str(),

			CI_PIPELINE_IID: num(),

			CI_MERGE_REQUEST_IID: num({ default: 0 }),
		});
	}

	return instance;
};

export { Env, getEnv };
