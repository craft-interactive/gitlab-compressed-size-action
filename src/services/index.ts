export * from "./env";
export * from "./gitlab-api";
export * from "./zip";
export * from "./diff";
export * from "./file";
export * from "./logger";
export * from "./http";
