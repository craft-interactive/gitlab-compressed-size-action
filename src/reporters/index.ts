import { gitlabPullRequestNote } from "./gitlab-pr-note";
import { stdout } from "./stdout";

export const reporters = {
  stdout,
  "gitlab-pr-note": gitlabPullRequestNote,
};
