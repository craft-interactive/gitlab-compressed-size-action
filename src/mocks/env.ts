import { getEnv, Env } from "../services";

jest.mock("../services/env");

const createEnvMockAPI = () => {
	const defaults: Env = {
		CI_API_V4_URL: "https://gitlab.com/api/v4",
		CI_JOB_NAME: "file-size-change",
		CI_PROJECT_ID: 2,
		CI_PIPELINE_IID: 50192,
		CI_MERGE_REQUEST_IID: 2501,
	};
	let mocked: Env = {
		...defaults,
	};

	return {
		mock: (overrides: Partial<Env>) => {
			mocked = {
				...defaults,
				...mocked,
				...overrides,
			};
		},
		setup: () => {
			(getEnv as jest.Mock).mockImplementation(() => mocked);
		},
		teardown: () => {
			mocked = {
				...defaults,
			};
			jest.unmock("../services/env");
		},
	};
};

export { createEnvMockAPI };
