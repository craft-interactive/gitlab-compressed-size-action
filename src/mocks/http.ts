import { http, HttpClientMethod, HttpClientOptions } from "../services";

type HttpRequestMock = {
	statusCode: number;
	body: any;
	url: string;
	method: HttpClientMethod;
};

const createHttpMockAPI = () => {
	let mocks: HttpRequestMock[] = [];

	return {
		mock: (mock: HttpRequestMock) => {
			mocks.push(mock);
		},
		setup: () => {
			// @ts-expect-error
			jest.spyOn(http, "createClient").mockImplementation(() => {
				const requestCountByMethodAndUrl: Record<string, number> = {};

				return async <Data = any>(url: string, options: HttpClientOptions) => {
					const id = `${options.method} ${url}`;
					const count = requestCountByMethodAndUrl[id] ?? 0;
					const matches = mocks.filter(
						(mock) => mock.url === url && mock.method === options.method
					);
					const match = matches[count] ?? matches[0];

					requestCountByMethodAndUrl[id]++;

					if (!match) {
						console.warn(
							`Unable to find http request mock for ${id}, did you forget to add the corresponding mock to your test suite?`
						);
					}

					return {
						statusCode: match?.statusCode ?? 200,
						body: (match?.body ?? null) as Data,
					};
				};
			});
		},
		teardown: () => {
			mocks = [];
		},
	};
};

export { createHttpMockAPI };
