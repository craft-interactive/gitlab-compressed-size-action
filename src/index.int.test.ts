import { createHttpMockAPI, createEnvMockAPI } from "./mocks";
import { zip } from "./services";
import { reporters } from "./reporters";
import { diff } from "./index";

jest.mock("consola");

describe("diff()", () => {
	const http = createHttpMockAPI();
	const env = createEnvMockAPI();

	beforeEach(async () => {
		Object.keys(reporters).forEach((id) => {
			jest.spyOn(reporters, id as any).mockImplementation(async () => {});
		});

		env.setup();
		http.setup();
	});

	afterEach(async () => {
		env.teardown();
		http.teardown();
	});

	it("should read the current fs.stats from disk and write them to the outFile path", async () => {
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/pipelines?ref=main",
			statusCode: 200,
			body: [],
		});

		const results = await diff({
			filePatterns: ["./src/mocks/files/*.css", "./src/mocks/files/*.zip"],
			outFile: ".tmp/.gitlab-compressed-size-action-report.json",
			auth: "my-gitlab-token",
			reporters: ["gitlab-pr-note"],
		});

		expect(results).toMatchSnapshot();
	});

	it("should retreive the previous fs.stats from the GitLab API and create a diff based on the current fs.stats from disk", async () => {
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/pipelines?ref=main",
			statusCode: 200,
			body: [{ id: 20391 }],
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/pipelines/20391/jobs?include_retried=true",
			statusCode: 200,
			body: [
				{ id: 219270, stage: "install", name: "install", status: "success" },
				{ id: 219271, stage: "build", name: "build", status: "success" },
				{
					id: 219273,
					stage: "build",
					name: "file-size-change",
					status: "success",
				},
				{ id: 219274, stage: "test", name: "test", status: "success" },
			],
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/jobs/219273/artifacts",
			statusCode: 200,
			body: await zip.create([
				{
					name: ".tmp/.gitlab-compressed-size-action-report.json",
					contents: JSON.stringify([
						{ path: "src/mocks/files/normalize.css", bytes: 6251 },
						{ path: "src/mocks/files/normalize.css.zip", bytes: 1941 },
					]),
				},
			]),
		});

		const results = await diff({
			filePatterns: ["./src/mocks/files/*.css", "./src/mocks/files/*.zip"],
			outFile: ".tmp/.gitlab-compressed-size-action-report.json",
			auth: "my-gitlab-token",
			reporters: ["gitlab-pr-note"],
		});

		expect(results).toMatchSnapshot();
	});

	it("should not fail if retrival of the previous fs.stats from the GitLab API failed", async () => {
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/pipelines?ref=main",
			statusCode: 500,
			body: "Whelp that's not good",
		});

		const results = await diff({
			filePatterns: ["./src/mocks/files/*.css", "./src/mocks/files/*.zip"],
			outFile: ".tmp/.gitlab-compressed-size-action-report.json",
			auth: "my-gitlab-token",
			reporters: ["gitlab-pr-note"],
		});

		expect(results).toMatchSnapshot();
	});

	it("should retry the retrival from the GitLab API if the last pipeline is still in 'created' state", async () => {
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/pipelines?ref=main",
			statusCode: 200,
			body: [{ id: 20391 }],
		});

		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/pipelines/20391/jobs?include_retried=true",
			statusCode: 200,
			body: [{ id: 219273 }],
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/jobs/219273",
			statusCode: 200,
			body: {
				id: 219273,
				stage: "build",
				name: "file-size-change",
				status: "created", // ! Simulate that the job is created upon the first retrival of the job.
			},
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/jobs/219273",
			statusCode: 200,
			body: {
				id: 219273,
				stage: "build",
				name: "file-size-change",
				status: "success",
			},
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/jobs/219273/artifacts",
			statusCode: 200,
			body: await zip.create([
				{
					name: ".tmp/.gitlab-compressed-size-action-report.json",
					contents: JSON.stringify([
						{ path: "src/mocks/files/normalize.css", bytes: 6251 },
						{ path: "src/mocks/files/normalize.css.zip", bytes: 1941 },
					]),
				},
			]),
		});

		const results = await diff({
			filePatterns: ["./src/mocks/files/*.css", "./src/mocks/files/*.zip"],
			outFile: ".tmp/.gitlab-compressed-size-action-report.json",
			auth: "my-gitlab-token",
			reporters: ["gitlab-pr-note"],
		});

		expect(results).toMatchSnapshot();
	});

	it("should retry the retrival from the GitLab API if the last pipeline is still in 'pending' state", async () => {
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/pipelines?ref=main",
			statusCode: 200,
			body: [{ id: 20391 }],
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/pipelines/20391/jobs?include_retried=true",
			statusCode: 200,
			body: [{ id: 219273 }],
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/jobs/219273",
			statusCode: 200,
			body: {
				id: 219273,
				stage: "build",
				name: "file-size-change",
				status: "pending", // ! Simulate that the job is pending upon the first retrival of the job.
			},
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/jobs/219273",
			statusCode: 200,
			body: {
				id: 219273,
				stage: "build",
				name: "file-size-change",
				status: "success",
			},
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/jobs/219273/artifacts",
			statusCode: 200,
			body: await zip.create([
				{
					name: ".tmp/.gitlab-compressed-size-action-report.json",
					contents: JSON.stringify([
						{ path: "src/mocks/files/normalize.css", bytes: 6251 },
						{ path: "src/mocks/files/normalize.css.zip", bytes: 1941 },
					]),
				},
			]),
		});

		const results = await diff({
			filePatterns: ["./src/mocks/files/*.css", "./src/mocks/files/*.zip"],
			outFile: ".tmp/.gitlab-compressed-size-action-report.json",
			auth: "my-gitlab-token",
			reporters: ["gitlab-pr-note"],
		});

		expect(results).toMatchSnapshot();
	});

	it("should retry the retrival from the GitLab API if the last pipeline is still in 'running' state", async () => {
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/pipelines?ref=main",
			statusCode: 200,
			body: [{ id: 20391 }],
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/pipelines/20391/jobs?include_retried=true",
			statusCode: 200,
			body: [{ id: 219273 }],
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/jobs/219273",
			statusCode: 200,
			body: {
				id: 219273,
				stage: "build",
				name: "file-size-change",
				status: "running", // ! Simulate that the job is running upon the first retrival of the job.
			},
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/jobs/219273",
			statusCode: 200,
			body: {
				id: 219273,
				stage: "build",
				name: "file-size-change",
				status: "success",
			},
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/jobs/219273/artifacts",
			statusCode: 200,
			body: await zip.create([
				{
					name: ".tmp/.gitlab-compressed-size-action-report.json",
					contents: JSON.stringify([
						{ path: "src/mocks/files/normalize.css", bytes: 6251 },
						{ path: "src/mocks/files/normalize.css.zip", bytes: 1941 },
					]),
				},
			]),
		});

		const results = await diff({
			filePatterns: ["./src/mocks/files/*.css", "./src/mocks/files/*.zip"],
			outFile: ".tmp/.gitlab-compressed-size-action-report.json",
			auth: "my-gitlab-token",
			reporters: ["gitlab-pr-note"],
		});

		expect(results).toMatchSnapshot();
	});

	it("should retry the retrival from the GitLab API if the last pipeline is still in 'waiting_for_resource' state", async () => {
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/pipelines?ref=main",
			statusCode: 200,
			body: [{ id: 20391 }],
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/pipelines/20391/jobs?include_retried=true",
			statusCode: 200,
			body: [{ id: 219273 }],
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/jobs/219273",
			statusCode: 200,
			body: {
				id: 219273,
				stage: "build",
				name: "file-size-change",
				status: "waiting_for_resource", // ! Simulate that the job is waiting_for_resource upon the first retrival of the job.
			},
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/jobs/219273",
			statusCode: 200,
			body: {
				id: 219273,
				stage: "build",
				name: "file-size-change",
				status: "success",
			},
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/jobs/219273/artifacts",
			statusCode: 200,
			body: await zip.create([
				{
					name: ".tmp/.gitlab-compressed-size-action-report.json",
					contents: JSON.stringify([
						{ path: "src/mocks/files/normalize.css", bytes: 6251 },
						{ path: "src/mocks/files/normalize.css.zip", bytes: 1941 },
					]),
				},
			]),
		});

		const results = await diff({
			filePatterns: ["./src/mocks/files/*.css", "./src/mocks/files/*.zip"],
			outFile: ".tmp/.gitlab-compressed-size-action-report.json",
			auth: "my-gitlab-token",
			reporters: ["gitlab-pr-note"],
		});

		expect(results).toMatchSnapshot();
	});

	it("should fail if an individual file failed the 'each' threshold", async () => {
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/pipelines?ref=main",
			statusCode: 200,
			body: [],
		});
		let error;

		try {
			await diff({
				filePatterns: ["./src/mocks/files/*.css", "./src/mocks/files/*.zip"],
				outFile: ".tmp/.gitlab-compressed-size-action-report.json",
				auth: "my-gitlab-token",
				reporters: ["gitlab-pr-note"],
				thresholds: {
					each: "6 KB",
				},
			});
		} catch (err) {
			error = err;
		}

		expect(error).toBeDefined();
		expect(error).toMatchObject(
			new Error(
				"One or more files do not meet the specified thresholds. See reporter output above for more details."
			)
		);
	});

	it("should fail if all files in summary failed the 'overall' threshold", async () => {
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/pipelines?ref=main",
			statusCode: 200,
			body: [],
		});
		let error;

		try {
			await diff({
				filePatterns: ["./src/mocks/files/*.css", "./src/mocks/files/*.zip"],
				outFile: ".tmp/.gitlab-compressed-size-action-report.json",
				auth: "my-gitlab-token",
				reporters: ["gitlab-pr-note"],
				thresholds: {
					overall: "7 KB",
				},
			});
		} catch (err) {
			error = err;
		}

		expect(error).toBeDefined();
		expect(error).toMatchObject(
			new Error(
				"One or more files do not meet the specified thresholds. See reporter output above for more details."
			)
		);
	});

	it("should execute the reporters and gracefully handle errors", async () => {
		jest.spyOn(reporters, "gitlab-pr-note").mockImplementation(async () => {
			throw new Error("nah");
		});
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/pipelines?ref=main",
			statusCode: 200,
			body: [],
		});
		let error;

		try {
			await diff({
				filePatterns: ["./src/mocks/files/*.css", "./src/mocks/files/*.zip"],
				outFile: ".tmp/.gitlab-compressed-size-action-report.json",
				auth: "my-gitlab-token",
				reporters: ["gitlab-pr-note"],
			});
		} catch (err) {
			error = err;
		}

		expect(reporters["gitlab-pr-note"]).toHaveBeenCalledTimes(1);
		expect(error).not.toBeDefined();
	});

	it("should gracefully handle invalid reporter IDs", async () => {
		http.mock({
			method: "GET",
			url: "https://gitlab.com/api/v4/projects/2/pipelines?ref=main",
			statusCode: 200,
			body: [],
		});
		let error;

		try {
			await diff({
				filePatterns: ["./src/mocks/files/*.css", "./src/mocks/files/*.zip"],
				outFile: ".tmp/.gitlab-compressed-size-action-report.json",
				auth: "my-gitlab-token",
				// @ts-expect-error
				reporters: ["invalid-reporter-id"],
			});
		} catch (err) {
			error = err;
		}

		expect(error).not.toBeDefined();
	});
});
