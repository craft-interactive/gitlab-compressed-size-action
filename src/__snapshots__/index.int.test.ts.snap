// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diff() should not fail if retrival of the previous fs.stats from the GitLab API failed 1`] = `
[
  {
    "change": {
      "percent": 100,
      "pretty": "+6.1 KB",
      "raw": 6251,
    },
    "id": "src/mocks/files/normalize.css",
    "isBelowThreshold": true,
    "size": {
      "pretty": "6.1 KB",
      "raw": 6251,
    },
    "status": "added",
  },
  {
    "change": {
      "percent": 100,
      "pretty": "+1.9 KB",
      "raw": 1941,
    },
    "id": "src/mocks/files/normalize.css.zip",
    "isBelowThreshold": true,
    "size": {
      "pretty": "1.9 KB",
      "raw": 1941,
    },
    "status": "added",
  },
  {
    "change": {
      "percent": 100,
      "pretty": "+8 KB",
      "raw": 8192,
    },
    "id": "summary",
    "isBelowThreshold": true,
    "size": {
      "pretty": "8 KB",
      "raw": 8192,
    },
    "status": "added",
  },
]
`;

exports[`diff() should read the current fs.stats from disk and write them to the outFile path 1`] = `
[
  {
    "change": {
      "percent": 100,
      "pretty": "+6.1 KB",
      "raw": 6251,
    },
    "id": "src/mocks/files/normalize.css",
    "isBelowThreshold": true,
    "size": {
      "pretty": "6.1 KB",
      "raw": 6251,
    },
    "status": "added",
  },
  {
    "change": {
      "percent": 100,
      "pretty": "+1.9 KB",
      "raw": 1941,
    },
    "id": "src/mocks/files/normalize.css.zip",
    "isBelowThreshold": true,
    "size": {
      "pretty": "1.9 KB",
      "raw": 1941,
    },
    "status": "added",
  },
  {
    "change": {
      "percent": 100,
      "pretty": "+8 KB",
      "raw": 8192,
    },
    "id": "summary",
    "isBelowThreshold": true,
    "size": {
      "pretty": "8 KB",
      "raw": 8192,
    },
    "status": "added",
  },
]
`;

exports[`diff() should retreive the previous fs.stats from the GitLab API and create a diff based on the current fs.stats from disk 1`] = `
[
  {
    "change": {
      "percent": 0,
      "pretty": "0 Bytes",
      "raw": 0,
    },
    "id": "src/mocks/files/normalize.css",
    "isBelowThreshold": true,
    "size": {
      "pretty": "6.1 KB",
      "raw": 6251,
    },
    "status": "unchanged",
  },
  {
    "change": {
      "percent": 0,
      "pretty": "0 Bytes",
      "raw": 0,
    },
    "id": "src/mocks/files/normalize.css.zip",
    "isBelowThreshold": true,
    "size": {
      "pretty": "1.9 KB",
      "raw": 1941,
    },
    "status": "unchanged",
  },
  {
    "change": {
      "percent": 0,
      "pretty": "0 Bytes",
      "raw": 0,
    },
    "id": "summary",
    "isBelowThreshold": true,
    "size": {
      "pretty": "8 KB",
      "raw": 8192,
    },
    "status": "unchanged",
  },
]
`;

exports[`diff() should retry the retrival from the GitLab API if the last pipeline is still in 'created' state 1`] = `
[
  {
    "change": {
      "percent": 100,
      "pretty": "+6.1 KB",
      "raw": 6251,
    },
    "id": "src/mocks/files/normalize.css",
    "isBelowThreshold": true,
    "size": {
      "pretty": "6.1 KB",
      "raw": 6251,
    },
    "status": "added",
  },
  {
    "change": {
      "percent": 100,
      "pretty": "+1.9 KB",
      "raw": 1941,
    },
    "id": "src/mocks/files/normalize.css.zip",
    "isBelowThreshold": true,
    "size": {
      "pretty": "1.9 KB",
      "raw": 1941,
    },
    "status": "added",
  },
  {
    "change": {
      "percent": 100,
      "pretty": "+8 KB",
      "raw": 8192,
    },
    "id": "summary",
    "isBelowThreshold": true,
    "size": {
      "pretty": "8 KB",
      "raw": 8192,
    },
    "status": "added",
  },
]
`;

exports[`diff() should retry the retrival from the GitLab API if the last pipeline is still in 'pending' state 1`] = `
[
  {
    "change": {
      "percent": 100,
      "pretty": "+6.1 KB",
      "raw": 6251,
    },
    "id": "src/mocks/files/normalize.css",
    "isBelowThreshold": true,
    "size": {
      "pretty": "6.1 KB",
      "raw": 6251,
    },
    "status": "added",
  },
  {
    "change": {
      "percent": 100,
      "pretty": "+1.9 KB",
      "raw": 1941,
    },
    "id": "src/mocks/files/normalize.css.zip",
    "isBelowThreshold": true,
    "size": {
      "pretty": "1.9 KB",
      "raw": 1941,
    },
    "status": "added",
  },
  {
    "change": {
      "percent": 100,
      "pretty": "+8 KB",
      "raw": 8192,
    },
    "id": "summary",
    "isBelowThreshold": true,
    "size": {
      "pretty": "8 KB",
      "raw": 8192,
    },
    "status": "added",
  },
]
`;

exports[`diff() should retry the retrival from the GitLab API if the last pipeline is still in 'running' state 1`] = `
[
  {
    "change": {
      "percent": 100,
      "pretty": "+6.1 KB",
      "raw": 6251,
    },
    "id": "src/mocks/files/normalize.css",
    "isBelowThreshold": true,
    "size": {
      "pretty": "6.1 KB",
      "raw": 6251,
    },
    "status": "added",
  },
  {
    "change": {
      "percent": 100,
      "pretty": "+1.9 KB",
      "raw": 1941,
    },
    "id": "src/mocks/files/normalize.css.zip",
    "isBelowThreshold": true,
    "size": {
      "pretty": "1.9 KB",
      "raw": 1941,
    },
    "status": "added",
  },
  {
    "change": {
      "percent": 100,
      "pretty": "+8 KB",
      "raw": 8192,
    },
    "id": "summary",
    "isBelowThreshold": true,
    "size": {
      "pretty": "8 KB",
      "raw": 8192,
    },
    "status": "added",
  },
]
`;

exports[`diff() should retry the retrival from the GitLab API if the last pipeline is still in 'waiting_for_resource' state 1`] = `
[
  {
    "change": {
      "percent": 100,
      "pretty": "+6.1 KB",
      "raw": 6251,
    },
    "id": "src/mocks/files/normalize.css",
    "isBelowThreshold": true,
    "size": {
      "pretty": "6.1 KB",
      "raw": 6251,
    },
    "status": "added",
  },
  {
    "change": {
      "percent": 100,
      "pretty": "+1.9 KB",
      "raw": 1941,
    },
    "id": "src/mocks/files/normalize.css.zip",
    "isBelowThreshold": true,
    "size": {
      "pretty": "1.9 KB",
      "raw": 1941,
    },
    "status": "added",
  },
  {
    "change": {
      "percent": 100,
      "pretty": "+8 KB",
      "raw": 8192,
    },
    "id": "summary",
    "isBelowThreshold": true,
    "size": {
      "pretty": "8 KB",
      "raw": 8192,
    },
    "status": "added",
  },
]
`;
